### PR TITLE
Observability

### PR DESCRIPTION
**Pull request by finn-block**
This resolves the issues outlined in #132

* Tracing:
  * configured from environment variable
  * spans have been added on all of the critical paths.
* Metrics:
  * All of the tracing spans also generate timing and count metrics, which is sufficient for now. No additional metrics have been added
* Logging:
  * All `Debugf`/`Infof`/`Warnf` logrus calls have been updated to use fields and a static log message

another addition not mentioned in that issue is a simple traffic generator in `impl/integrationtest/main.go` which generates a DID, PUTs it and GETs it every 30 seconds.

Sample PUT span:

image

### CODE REVIEW

File: impl/pkg/dht/dht_test.go
Comment (thereisnogabe) [inline impl/pkg/dht/dht_test.go]: is there a `m.Fail()` equivalent so we don't need to panic?

also should the contexts here and below be the same?

File: impl/pkg/server/server_test.go
Comment (thereisnogabe) [inline impl/pkg/server/server_test.go]: same comment here

File: impl/pkg/telemetry/telemetry.go
Comment (thereisnogabe) [inline impl/pkg/telemetry/telemetry.go]: interesting...how does this work?

we do/should have a version set in our config

File: impl/trafficgen/main.go
Comment (thereisnogabe) [inline impl/trafficgen/main.go]: is this intended to be an integration test?

may be worth renaming the package as such

File: impl/pkg/dht/dht_test.go
Comment (finn-block) [reply impl/pkg/dht/dht_test.go]: not seeing an `m.Fail()`, just `m.Run()` (https://pkg.go.dev/testing#M). Context doesn't matter, these TestMains are just required to make sure the tracer isn't null when the test runs.

File: impl/pkg/telemetry/telemetry.go
Comment (finn-block) [reply impl/pkg/telemetry/telemetry.go]: Read about it [here](https://pkg.go.dev/runtime/debug#ReadBuildInfo), right now it's just returning `(devel)` for the current module's version:

Screenshot 2024-03-05 at 10 27 05 AM

I thought that would become a useful name when we add tags, but I just tried adding a tag and it didn't help, so I might try to push that forward a little before merging this.

File: impl/trafficgen/main.go
Comment (finn-block) [reply impl/trafficgen/main.go]: yeah probably. I'll rename the package.

File: impl/pkg/telemetry/telemetry.go
Comment (finn-block) [reply impl/pkg/telemetry/telemetry.go]: Okay so i thought i had successfully used this in the past but i guess [it will never work](https://github.com/golang/go/issues/29228). Looking at where I thought I had successfully used it, I appear to have some compile-time variable setting to inject a version externally. Will add the same here.

Comment (finn-block) [reply impl/pkg/telemetry/telemetry.go]: image

looks slightly better now. GitHub [doesn't give us a lot of great options](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context), and because the Dockerfile copies the `impl/` not the git root, we can't use `git describe` or anything like that in the container. For tagged releases, the string `telemetry` in that screenshot should be the tag name.

File: impl/pkg/telemetry/telemetry.go
Comment (thereisnogabe) [inline impl/pkg/telemetry/telemetry.go]: hmm, don't love the idea of having a global tracer

can we put this on a struct which `SetupTelemetry` returns?

Comment (finn-block) [reply impl/pkg/telemetry/telemetry.go]: Totally could. Is there a problem with having a global tracer? An alternative to having a tracer that gets passed around everywhere is to initialize a tracer in each package that we can use to instrument that package, which is what the gin middleware does.

Comment (thereisnogabe) [reply impl/pkg/telemetry/telemetry.go]: Global state scares me generally because of assumptions it caries - may/not be initialized, different callers using it for different purposes in an unsafe manner. Of course this is negligible given how small the codebase is now, though I view it as a best practice to avoid globally mutable state

Comment (finn-block) [reply impl/pkg/telemetry/telemetry.go]: Reasonable. I'll figure out how I want to setup non-global tracer(s)

**Review by thereisnogabe — APPROVED** [review]
approving pending global state change

nice work 🎉

File: impl/pkg/telemetry/telemetry.go
Comment (finn-block) [reply impl/pkg/telemetry/telemetry.go]: So I refactored a bit to have a `GetTracer()` function that returns a tracer. There's still one instance that gets initialized on first call, which isnt great but I'm not sure what better options there are. I could make a function + tracer instance like this for each package (then there wouldnt be a global one technically) but I dont see how that's any better. Let me know if you this this is acceptable, if not let's discuss options.

Comment (thereisnogabe) [reply impl/pkg/telemetry/telemetry.go]: I remember going through a similar type of conversation with @andresuribe87 on something a while back and he convinced me to rip out all the (complex, ugly) singleton code and just make a struct that returns a unique thing, which worked for the 90% case.

I believe what you've done here is fine too, and we should be open to changing it should that assumption be false

Comment (finn-block) [reply impl/pkg/telemetry/telemetry.go]: I could see doing that too. It would involve passing it around through a lot of different functions, which seems annoying. I'm going to merge this as is, but i'm not opposed to moving to that in the future.

**Pull request merged**
